core/primitives/src/errors.rs

```diff
@@ -26,74 +77,182 @@ impl std::fmt::Display for StorageError {
+    NotEnoughAllowance {
+        account_id: AccountId,
+        public_key: PublicKey,
+        #[serde(with = "u128_dec_format")]
+        allowance: Balance,
+        #[serde(with = "u128_dec_format")]
+        cost: Balance,
+    },
+}
+
+#[derive(
+    BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize, RpcError,
+)]
+#[rpc_error_variant = "ActionError"]
+pub struct ActionError {
+    pub index: Option<u64>,
+    pub kind: ActionErrorKind,
 }
 
-#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
-pub enum ActionError {
-    AccountAlreadyExists(AccountId),
-    AccountDoesNotExist(String, AccountId),
-    CreateAccountNotAllowed(AccountId, AccountId),
-    ActorNoPermission(AccountId, AccountId, String),
-    DeleteKeyDoesNotExist(AccountId),
-    AddKeyAlreadyExists(PublicKey),
-    DeleteAccountStaking(AccountId),
-    DeleteAccountHasRent(AccountId, Balance),
-    RentUnpaid(AccountId, Balance),
-    TriesToUnstake(AccountId),
-    TriesToStake(AccountId, Balance, Balance, Balance),
-    FunctionCallError(String), // TODO type
+#[derive(
+    BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize, RpcError,
+)]
+#[rpc_error_variant = "ActionError"]
+pub enum ActionErrorKind {
+    /// Happens when CreateAccount action tries to create an account with account_id which is already exists in the storage
+    AccountAlreadyExists {
+        account_id: AccountId,
+    },
+    /// Happens when TX receiver_id doesn't exist (but action is not Action::CreateAccount)
+    AccountDoesNotExist {
+        account_id: AccountId,
+    },
+    /// A newly created account must be under a namespace of the creator account
+    CreateAccountNotAllowed {
+        account_id: AccountId,
+        predecessor_id: AccountId,
+    },
+    /// Administrative actions like DeployContract, Stake, AddKey, DeleteKey. can be proceed only if sender=receiver.
+    ActorNoPermission {
+        account_id: AccountId,
+        actor_id: AccountId,
+    },
+    /// Account tries to remove an access key that doesn't exist
+    DeleteKeyDoesNotExist {
+        account_id: AccountId,
+        public_key: PublicKey,
+    },
+    /// The public key is already used for an existing access key
+    AddKeyAlreadyExists {
+        account_id: AccountId,
+        public_key: PublicKey,
+    },
+    /// Account is staking and can not be deleted
+    DeleteAccountStaking {
+        account_id: AccountId,
+    },
+    /// Have to cover rent before delete an account
+    DeleteAccountHasRent {
+        account_id: AccountId,
+        #[serde(with = "u128_dec_format")]
+        balance: Balance,
+    },
+    /// ActionReceipt can't be applied for an account. Account state rent is not payed.
+    RentUnpaid {
+        /// An account which is required to pay the rent
+        account_id: AccountId,
+        /// Required balance to cover the state rent
```

Comment:
This is still confusing.
```suggestion
        /// Rent due to pay.
```